src/tracks/playabletrack/wavetrack/ui/SpectrumCache.cpp

```diff
@@ -22,33 +22,33 @@
 namespace {
 
 static void ComputeSpectrumUsingRealFFTf(PffftFloats buf,
-   const FFTParam *hFFT,
-   const float * __restrict window, size_t len, PffftFloats output)
+   PFFFT_Setup &setup,
+   const float * __restrict window, size_t len,
+   PffftFloats work, PffftFloats output)
 {
    const auto buffer = buf.get();
    const auto out = output.get();
-   size_t i;
-   if(len > hFFT->Points * 2)
-      len = hFFT->Points * 2;
-   for(i = 0; i < len; i++)
-      buffer[i] *= window[i];
-   for( ; i < (hFFT->Points * 2); i++)
-      buffer[i] = 0; // zero pad as needed
-   RealFFTf(buffer, hFFT);
+   {
+      size_t i;
+      for (i = 0; i < len; ++i)
+         buffer[i] *= window[i];
+      for( ; i < len; ++i)
+         buffer[i] = 0; // zero pad as needed
+   }
+   pffft_transform_ordered(&setup, buffer, buffer, work.get(), PFFFT_FORWARD);
    // Handle the (real-only) DC
    float power = buffer[0] * buffer[0];
    if(power <= 0)
       out[0] = -160.0;
    else
       out[0] = 10.0 * log10f(power);
-   for(i = 1; i < hFFT->Points; i++) {
-      const int index = hFFT->BitReversed[i];
+   for (size_t index = 2; index < len; index += 2) {
       const float re = buffer[index], im = buffer[index + 1];
       power = re * re + im * im;
       if(power <= 0)
-         out[i] = -160.0;
+         out[index / 2] = -160.0;
       else
-         out[i] = 10.0*log10f(power);
+         out[index / 2] = 10.0 * log10f(power);
```

Comment:
Hey, maybe we could replace this log10f with an optimized log2. We saw in the tempo-estimation optimization PR that it sped up the STFT a great deal. (We probably should not blindly reuse the `GetLog2` of `MirDsp.cpp`, though, because that approximation is only accurate to two decimal places, as you correctly found out, @Paul-Licameli ).

Reply:
No.

Don’t use approximation for speed. Plot the best spectral data we can calculate.